File: mcp_observatory/cost/pricing.py
"""Pricing and cost estimation utilities for model invocations."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict


@dataclass(frozen=True)
class ModelPricing:
    """Per-token pricing for a model in USD."""

    input_per_1k: float
    output_per_1k: float


DEFAULT_PRICING = ModelPricing(input_per_1k=0.0020, output_per_1k=0.0040)
MODEL_PRICING: Dict[str, ModelPricing] = {
    "gpt-4o-mini": ModelPricing(input_per_1k=0.00015, output_per_1k=0.00060),
    "gpt-4.1-mini": ModelPricing(input_per_1k=0.00040, output_per_1k=0.00160),
    "claude-3-5-sonnet": ModelPricing(input_per_1k=0.00300, output_per_1k=0.01500),
}


def get_pricing(model: str) -> ModelPricing:
    """Get pricing for a model name, falling back to a default."""
    return MODEL_PRICING.get(model, DEFAULT_PRICING)


def estimate_cost(model: str, request_tokens: int, response_tokens: int) -> float:
    """Estimate total USD cost from token usage."""
    pricing = get_pricing(model)
    input_cost = (max(0, request_tokens) / 1000.0) * pricing.input_per_1k
    output_cost = (max(0, response_tokens) / 1000.0) * pricing.output_per_1k
    return round(input_cost + output_cost, 8)

File: src/cost/pricing.ts
const PRICING: Record<string, { input: number; output: number }> = {
  'gpt-4': { input: 0.03, output: 0.06 },
  'gpt-4-turbo': { input: 0.01, output: 0.03 },
  'gpt-4o': { input: 0.005, output: 0.015 },
  'gpt-4o-mini': { input: 0.00015, output: 0.0006 },
  'claude-3-opus': { input: 0.015, output: 0.075 },
  'claude-3-sonnet': { input: 0.003, output: 0.015 },
  'claude-3-haiku': { input: 0.00025, output: 0.00125 },
};

export function estimateCost(
  totalTokens: number,
  model: string = 'gpt-4',
  source: 'agent' | 'model' = 'model'
): number {
  const pricing = PRICING[model] || PRICING['gpt-4'];

  const inputTokens = Math.ceil(totalTokens * 0.5);
  const outputTokens = totalTokens - inputTokens;

  const inputCost = (inputTokens / 1000) * pricing.input;
  const outputCost = (outputTokens / 1000) * pricing.output;

  return inputCost + outputCost;
}

export function getPricing(model: string): { input: number; output: number } {
  return PRICING[model] || PRICING['gpt-4'];
}
